### ISSUE TITLE
New `cmd` or `exec` source

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#992]
After talking with @lukesteensen he had an interesting idea of using a command as data input. This command would be executed on an interval, the output would be ingested as either log or metrics events. This is an interesting approach to "plugins" in that users could write simple scripts that interface with any data source.

Before we begin work, we should outline scenarios this would support and how the incoming data should be structured, if at all. And would this support metrics?

**Comment by a-rodin** [comment]
I really like this idea.

Actually, there is a way to do this right now without a special source using a combination of a shell script, `netcat`, and `tcp` source:

`script.sh`:
```bash
#!/bin/bash
set -ex

(while true; do "my-command"; sleep 1; done) |
  nc 127.0.0.1 55555
```
`vector.toml`:
```toml
[sources.service]
  type = "tcp"
  address = "127.0.0.1:55555"
```

However, it is not very ergonomic and it is better to have it built in.

> This command would be executed on an interval, the output would be ingested ...

I think we could avoid forcing users to generate only one event per invocation (and thus each time necessarily create new process). The command could generate a stream of newline-delimited events until it exits. To improve reliability and flexibility we can add two additional parameter `exit_restart_interval` and which would specify the interval since the end of the previous invocation after which the command will be launched again. With these two parameters, the `command` itself and `restart_interval`, it would be possible to support both one-off commands and long-running stream generators.

Examples:

* Report size of a directory:
    ```toml
    [sources.dir_size_watcher]
    command = "du -h | cut -f1"
    restart_interval = 10 # seconds
    ```
    The command would be executed each 10 seconds.
* Make a stateful counter:
    ```toml
    [sources.stateful_counter]
    command = "./counter.sh"
    ```
    and the `counter.sh` would be
    ```sh
    #!/bin/sh
    for ((i = 0;; i++)); do
        echo $i
        sleep 1
    fi
    ```
* It should be possible to generate structured events this way too:
    ```toml
    [sources.structured_counter]
    command = "node ./structured_counter.js"
    encoding = "json"
    ```
    with `structured_counter.js` being a Node.js script:
    ```javascript
    for (let i = 0;; i++) {
        const event = {
            message: i.toString(),
            timestamp: new Date().toISOString()
        }
        console.log(JSON.stringify(event))
    }
    ```

The source could also be written in a high-performant language, such as Rust or C. 

Actually, before `kafka` source was implemented, I was trying to use [kafkacat](https://github.com/edenhill/kafkacat) combined with `stdin` source to do this. However, I encountered two issues:

1. `kafkacat` sometimes exited with an error, but nothing restarted it automatically
2. graceful shutdown was not possible because termination of Vector and `kafkacat` was not synchronized

Both of the problems mentioned above could have been solved if there were `cmd` sink from my examples above.

Also, the same approach can be used to create `cmd` transform and sink (need to be put into separate issues). Combined, `cmd` sources/transforms/sinks might even be a viable alternative to scripting from #1041, as they could provide the same benefits and not force the users to learn a specific programming language.

**Comment by lukesteensen** [comment]
@a-rodin That all sounds good! The one thing I'm a little hesitant about is encouraging long-running processes. Managing those correctly is non-trivial and not one of vector's core competencies. In the past, I've advocated for running those processes under a real process manager, connecting their stdout to a named pipe, and using the file source to ingest from the pipe.

**Comment by a-rodin** [comment]
@lukesteensen I like the approach with named pipes, it is robust and already allows to implement custom sources/transforms/sinks.

I have some concerns about portability. The development and production environments need to have the same process manager, also I'm not sure is it possible to use `file` source/sink this way on Windows if we port Vector to it.

But then, as an alternative, we can expose APIs similar to [child_process](https://nodejs.org/api/child_process.html) as part of adding scripting to Vector and allow users to manage long-running processes themselves if they want to.

**Comment by lukesteensen** [comment]
Portability is a good point. I'm not sure how common deployments to Windows will be, as we don't officially support them right now.

I'm open to considering a way to manage long-running processes, I just think we're better off starting with the simpler case. It's a complex thing to implement correctly, so I want to make sure the benefit is worth the cost of writing it, maintaining it, etc.

**Comment by jamtur01** [comment]
The ability to execute a script based collector on a cron-like schedule.

* Anything sent to stdout by the script should be read as event data (line separated).
* Anything send to stderr by the script should be internal logging to vector.

The following exit codes should be interpreted by Vector:

```
	0: successful completion
	1: error in execution (vector should log stderr to internal logging)
```

* Should have a limit for the execution time (the script author should be aware of that limit and has the responsibility to fail gracefully before timeout is reached).
* Should also (optionally) prevent concurrent execution (i.e. 1 minute schedule but 5 minute timeout).
* Should have options to nice and ionice the script being executed, where permissions allow.
* Should be able to set directory for execution

**Comment by jszwedko** [comment]
User requested this in discord: https://discordapp.com/channels/742820443487993987/746070591097798688/768207458258649098

**Comment by jszwedko** [comment]
Closed by #6876
